docs/yaml/functions/shared_library.yaml

```diff
@@ -44,3 +44,8 @@ kwargs:
     description: |
       Specify a Microsoft module definition file for controlling symbol exports,
       etc., on platforms where that is possible (e.g. Windows).
+
+  <lang>_shared_args:
+    type: str | file
```

Comment:
Why `file`?

Reply:
Checking how `<lang>_args` works, some parts of the code seems to allow File, but documentation does not mention that, and I see many code paths that does not seems to support it at first glance.

Allowing File does make sense, I guess it's common to pass a file in compiler arguments, but doc needs to be updated and it would have to be catched as new feature by KwargInfo. It would also need to ensure we have a unit test for that, because from reading the code I'm not convinced it works. Probably we need a convertor in KwargInfo to only get strings?

Reply:
I guess that should actually be `list[str | file]`. And that the documentation for `<lang>_args` is wrong too, since it says `list[str]`, but should also be `list[str | file]`

Reply:
I would leave this out of this PR because I'm not convinced it actually works, or add a unit test with a File in c_args to convince me...

If I understand correctly, we use that there: https://github.com/mesonbuild/meson/blob/master/mesonbuild/backend/ninjabackend.py#L2923, and will crash there: https://github.com/mesonbuild/meson/blob/master/mesonbuild/backend/backends.py#L959. But I may be missing something in how that code works.

Reply:
ahah, yes, as predicted:
```
  File "/home/xclaesse/programmation/meson/mesonbuild/backend/ninjabackend.py", line 2923, in _generate_single_compile_target_args
    commands += self.escape_extra_args(target.get_extra_args(compiler.get_language()))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xclaesse/programmation/meson/mesonbuild/backend/backends.py", line 959, in escape_extra_args
    if arg.startswith(('-D', '/D')):
       ^^^^^^^^^^^^^^
AttributeError: 'File' object has no attribute 'startswith'
```

Reply:
So this is extra awesome, because this totally works for Vala. So it has to be handled here, since we have to allow it or we'd regress working code...

There is even a `manual tests/7 vala composite widgets` that tests this, though it doesn't seem we have an automated test that does. So we can't take this out, it has to remain. Sigh

Reply:
But maybe we can handle this in the interpreter instead of sending it to the backend at all....

Reply:
Okay. So I did some git archeology. This was explicitly turned on for Vala in 0.25 (see: cd40187d4ecc99a91aaaf2cad8991d600691c443). That means that turning it off would constitute a regression in my mind, since the explicit purpose of the patch was to allow this. So, I've allowed this for Vala, and only for Vala. And I've moved the conversion code to the Interpreter level, so that the backend/middle layer doesn't have to be aware of this (except for creating the dependency)